src/wp-includes/l10n.php

```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
+	// Ensures the correct locale is set as the current one, in case it was filtered.
+	WP_I18n_Translation_Controller::instance()->set_locale( $locale );
 
-		return false;
+	/**
+	 * Filters the preferred file format for translation files.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $preferred_format Preferred file format. Possible values: 'php', 'mo'. Default: 'php'.
+	 */
+	$preferred_format = apply_filters( 'translation_file_format', 'php' );
+	if ( ! in_array( $preferred_format, array( 'php', 'mo' ), true ) ) {
+		$preferred_format = 'php';
 	}
 
-	if ( isset( $l10n[ $domain ] ) ) {
-		$mo->merge_with( $l10n[ $domain ] );
+	$mofile_preferred = str_replace( '.mo', ".mo.$preferred_format", $mofile );
+
+	if ( 'mo' !== $preferred_format || str_ends_with( $mofile, $preferred_format ) ) {
+
+		/** This filter is documented in wp-includes/l10n.php */
+		$mofile_preferred = apply_filters( 'load_textdomain_mofile', $mofile_preferred, $domain );
+
+		/**
+		 * Filters the file path for loading translations for the given text domain.
+		 *
+		 * The file could be an MO or PHP file.
+		 *
+		 * @since 6.5.0
+		 *
+		 * @param string $file   Path to the translation file to load.
+		 * @param string $domain The text domain.
+		 */
+		$mofile_preferred = apply_filters( 'load_translation_file', $mofile_preferred, $domain );
+
+		$success = WP_I18n_Translation_Controller::instance()->load( $mofile_preferred, $domain, $locale );
+
+		if ( $success ) {
+			if ( isset( $l10n[ $domain ] ) && $l10n[ $domain ] instanceof MO ) {
+				WP_I18n_Translation_Controller::instance()->load( $l10n[ $domain ]->get_filename(), $domain, $locale );
+			}
+
+			// Unset Noop_Translations reference in get_translations_for_domain.
+			unset( $l10n[ $domain ] );
+
+			$l10n[ $domain ] = new WP_I18n_Translations( $domain );
+
+			$wp_textdomain_registry->set( $domain, $locale, dirname( $mofile ) );
+
+			return true;
+		}
 	}
 
-	unset( $l10n_unloaded[ $domain ] );
+	/** This action is documented in wp-includes/l10n.php */
+	do_action( 'load_textdomain', $domain, $mofile );
+
+	/** This filter is documented in wp-includes/l10n.php */
+	$mofile = apply_filters( 'load_textdomain_mofile', $mofile, $domain );
+
+	/** This filter is documented in wp-includes/l10n.php */
+	$mofile = apply_filters( 'load_translation_file', $mofile, $domain );
+
+	$success = WP_I18n_Translation_Controller::instance()->load( $mofile, $domain, $locale );
+
+	if ( $success ) {
+		if ( isset( $l10n[ $domain ] ) && $l10n[ $domain ] instanceof MO ) {
+			WP_I18n_Translation_Controller::instance()->load( $l10n[ $domain ]->get_filename(), $domain, $locale );
+		}
 
-	$l10n[ $domain ] = &$mo;
+		// Unset Noop_Translations reference in get_translations_for_domain.
+		unset( $l10n[ $domain ] );
 
-	$wp_textdomain_registry->set( $domain, $locale, dirname( $mofile ) );
+		$l10n[ $domain ] = new WP_I18n_Translations( $domain );
+
+		$wp_textdomain_registry->set( $domain, $locale, dirname( $mofile ) );
+	}
```

Comment:
Most of this looks very much like duplicate code of what is in the if clause, with some tiny adjustments like probably the variables. Can we move this to a separate function (`@access private` would be okay) to avoid that?

Reply:
It's only a handful of lines, not really worth creating another function for this IMHO.

Reply:
Reviewing the code more closely, I don't even think we need a new function, we just need to simplify the code. I don't see a good reason for all these lines being duplicated. As far as I can tell, they are _exactly_ the same, just acting on a different file. We can probably refactor with a `foreach` loop to avoid the duplicate code. I'll post a suggestion.

Reply:
I just realized I can't leave a suggestion on that code because GH doesn't allow it, so just sharing here:
* We can use an array of files to attempt and then avoid this duplicate code by using a `foreach` loop.
* We run the logic for every file in the array (either the PHP file and then the MO file, or just the MO file), and as soon as one file was loaded we return `true`. Or, if for some reason nothing is loaded, we return `true` at the end, which would be the same as today.

Something like:
```php
$translation_files = array( $mofile );
if ( 'mo' !== $preferred_format ) {
	array_unshift(
		$translation_files,
		str_replace( '.mo', ".mo.$preferred_format", $mofile )
	);
}
foreach ( $translation_files as $translation_file ) {
	// All logic currently duplicated, using $translation_file instead of $mofile / $mofile_preferred.
	// Including `return true` on success.
}

// Only relevant if no file was actually loaded, in which case today the function also still returns true.
return true;
```